### ISSUE TITLE
DataLoader is slow

### ISSUE THREAD

**Issue by LeeHuan18** [rusty1s/pytorch_geometric#802]
Hi,
    I generate my own graphs and I follow the webpage instruct for the DataLoader.

```
data_list = [Data(...), ..., Data(...)]
loader = DataLoader(data_list, batch_size=32)
```

    But with this loader, the model trains very slow and the gpu consumption is only 2%. I doubt if it's because DataLoader runs slow?

    So I use a simple dnn and use this DataLoader, it's still slow and the gpu usage is still 2%.

```
class Net(torch.nn.Module):
    def __init__(self, in_features=7, num_classes=2):
        super(Net, self).__init__()
        self.fc = nn.Sequential(
                  nn.Linear(in_features*4, 2**8),
                  nn.BatchNorm1d(2**8),
                  nn.ReLU(),
                  nn.Linear(2**8, num_classes)
        )

    def forward(self, data):
        x, edge_index = data.x, data.edge_index
        x = x.view(-1, 28)
        x = self.fc(x)        
        return F.softmax(x, dim=1)

```

  So is it a problem of DataLoader or any possible reason?

**Comment by rusty1s** [comment]
I never witnessed this myself and always thought the `DataLoader` is fast as f***. Since your list is already holding `data` objects, this could only be caused by `Batch.from_data_list`. Can you actually measure the time consumption for a call such as:
```python
Batch.from_data_list(data_list[:32])
```
In addition, it would be great if you could send some details about your `data` objects. How big/small are your graphs?

**Comment by LeeHuan18** [comment]
@rusty1s 
Using `%%time`, the result is:

```
CPU times: user 6.99 ms, sys: 153 µs, total: 7.15 ms
Wall time: 22.9 ms
Batch(batch=[128], edge_index=[2, 0], x=[128, 7], y=[32])

```

I don't have a concept it's fast or not...

About the graph, each single graph, as you can see above, has 0-edge, 4-node and each node is a 1-D array with 7 features. Also there is a additional feature `y` as the label of the graph. So the graph is very small.

**Comment by rusty1s** [comment]
I think it is quite understandable that your gpu usage is low with such small graphs and a batch_size of 32.

**Comment by LeeHuan18** [comment]
So I increase the batch_size to 512, but it doesn't improve. Can you please give a hint how to improve it?

**Comment by SaschaStenger** [comment]
I asked a question, that might be related to yours. I had a high CPU load and a low GPU load and was able to fix it by changing the number of workers for the `data loader`
[Issue](https://github.com/rusty1s/pytorch_geometric/issues/811)

**Comment by LeeHuan18** [comment]
@SaschaStenger Thank you very much! let me try.

**Comment by ghorbanimahdi73** [comment]
Hi. I have a graph data loader class that overrides the __getitem__ with each datapoint having one anchor graph and 10 randomly selected graphs from the dataset and their label distances in a dictionary. During training the Dataloader becomes super slow relative to the number of graph neighbors (10 here) Following is a snapshot of the code. For making the dataset of neighbors I use the get() function to make a list of Data objects. Is there a way around this without using get()

<img width="748" alt="Screen Shot 2022-09-09 at 12 52 28 PM" src="https://user-images.githubusercontent.com/45661178/189433114-5080ae01-f397-46fb-b672-ffdbfb953a27.png">

**Comment by rusty1s** [comment]
Your code looks good to me. Do you know which is the slow part: getting all the `data` objects within `__getitem__` or collating them together during `DataLoader`?

**Comment by ghorbanimahdi73** [comment]
It is the `__getitem__` that makes it slow. Since every datapoint has 10 neighbors, the dataloading is 10 times slower if I also save a list of neighbors. I am using the get() function to retrieve each data objects. I tried using `index_select(idx)` and giving it an array of indices as idx but that would return the full dataset and not the selected indices as calling dataset.index_select(idx).data would show all the data points.

**Comment by rusty1s** [comment]
Since you are requesting 10 more objects, it is expected that the data access takes 10 times longer - I don‘t think there is much we can do here besides improving speed of `__getitem__` and batching overall, or by leveraging more workers. I still think that the main bottleneck should be in the collate function though (which does the heavy work of merging these objects into one), so might be good to confirm that this is indeed the case.

**Comment by ghorbanimahdi73** [comment]
So if I understand it correctly collate only affects the preprocessing of data and collating it into a huge Data object. Does it also affect batching during the data loading? In that case do I need to redefine the collate function?

**Comment by rusty1s** [comment]
So the `DataLoader` does two things: (1) It gets all graphs/examples that we want to batch together (the `__getitem__` call), (2) it batches these graphs together into a huge `data` object. Usually, step (2) is more expensive since (1) just refers to a simple list access without any real computation taking place.

**Comment by ghorbanimahdi73** [comment]
Sure, that is right. Collating takes time. In this case, should I avoid collating and load individual graphs from gpu scratch directory or is there another way to redefine the collate function?

**Comment by rusty1s** [comment]
Can you clarify what you mean with re-defining the collate function? I can look into any bottlenecks of the `collate` function. Do you have a small example to reproduce the bottleneck?

**Comment by rusty1s** [comment]
Do you have a reproducible example? What `num_workers` configuration are you using?
